extensions/amp-story/1.0/amp-story.js

```diff
@@ -1285,15 +1302,30 @@ export class AmpStory extends AMP.BaseElement {
 
     if (
       isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      targetPage.isPaywallProtected() &&
+      !this.granted_
     ) {
       if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
+        // Do nothing when the subscription dialog is already triggered.
         return Promise.resolve();
       }
       this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+      this.subscriptionsService_.selectAndActivatePlatform();
+      this.subscriptionsService_.addOnEntitlementResolvedCallback((e) => {
+        const {entitlement} = e;
+        if (
+          this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE) &&
+          entitlement.granted
+        ) {
+          this.granted_ = true;
+          this.storeService_.dispatch(
+            Action.TOGGLE_SUBSCRIPTIONS_DIALOG,
+            false
+          );
+        }
+      });
```

Comment:
This does not belong in the `switchTo`, please move it to where is appropriate

Reply:
Combining from the comment below, I think you mean only keep the logic that toggle the SUBSCRIPTIONS_DIALOG_STATE and have a listener maybe in amp-story-subscriptions extension to do all the detailed work to clean up states and trigger/dismiss the paywall, e.g. all of the SubscriptionService calls. 

For that one extra thing we would need to do is to record grantStatus so that it can be still read in amp-story. We can just add a new StateProperty like 'SUBSCRIPTIONS_GRANTED_STATUS'. Will do that later and feel free to correct me if I totally misunderstand you.

Reply:
Done.